sources/Valkey.Glide/PubSubMessageQueue.cs

```diff
@@ -0,0 +1,173 @@
+    /// <returns>true if a message was retrieved; otherwise, false.</returns>
+    /// <exception cref="ObjectDisposedException">Thrown when the queue has been disposed.</exception>
+    public bool TryGetMessage(out PubSubMessage? message)
+    {
+        ThrowIfDisposed();
+
+        if (_messages.TryDequeue(out message))
+        {
+            // Consume one semaphore count since we dequeued a message
+            _ = _messageAvailable.Wait(0);
+            return true;
+        }
+
+        message = null;
+        return false;
+    }
+
+    /// <summary>
+    /// Asynchronously wait for and retrieve a message from the queue.
+    /// </summary>
+    /// <param name="cancellationToken">Token to cancel the operation.</param>
+    /// <returns>A task that represents the asynchronous operation. The task result contains the retrieved message.</returns>
+    /// <exception cref="ObjectDisposedException">Thrown when the queue has been disposed.</exception>
+    /// <exception cref="OperationCanceledException">Thrown when the operation is cancelled.</exception>
+    public async Task<PubSubMessage> GetMessageAsync(CancellationToken cancellationToken = default)
+    {
+        ThrowIfDisposed();
+
+        // Wait for a message to be available
+        await _messageAvailable.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+        // Check if disposed after waiting
+        ThrowIfDisposed();
+
+        // Try to dequeue the message
+        if (_messages.TryDequeue(out PubSubMessage? message))
+        {
+            return message;
+        }
+
+        // This should not happen under normal circumstances, but handle it gracefully
+        throw new InvalidOperationException("Message queue is in an inconsistent state");
+    }
+
+    /// <summary>
+    /// Get an async enumerable for continuous message processing.
+    /// </summary>
+    /// <param name="cancellationToken">Token to cancel the enumeration.</param>
+    /// <returns>An async enumerable that yields messages as they become available.</returns>
+    /// <exception cref="ObjectDisposedException">Thrown when the queue has been disposed.</exception>
+    public async IAsyncEnumerable<PubSubMessage> GetMessagesAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        while (!_disposed && !cancellationToken.IsCancellationRequested)
+        {
+            PubSubMessage message;
+            try
+            {
+                message = await GetMessageAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (ObjectDisposedException)
+            {
+                // Queue was disposed, exit enumeration
+                yield break;
+            }
+            catch (OperationCanceledException)
+            {
+                // Operation was cancelled, exit enumeration
+                yield break;
+            }
+
+            yield return message;
+        }
+    }
```

Comment:
Not much familiarity with this domain, but...

Since `GetMessageAsync` already handles the case when the queue has already been disposed or the operation was cancelled, can we simplify this loop? Are we wanting to avoid the small overhead of raising/catching exceptions for performance benefits?


```suggestion
    public async IAsyncEnumerable<PubSubMessage> GetMessagesAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        while (true)
        {
            PubSubMessage message;
            try
            {
                message = await GetMessageAsync(cancellationToken).ConfigureAwait(false);
            }
            catch (ObjectDisposedException)
            {
                // Queue was disposed, exit enumeration
                yield break;
            }
            catch (OperationCanceledException)
            {
                // Operation was cancelled, exit enumeration
                yield break;
            }

            yield return message;
        }
    }
```

Reply:
We should not remove the while loop in this method. This is what allows us to process all messages rather than just a single message. When the GetMessages method is called, we will stay in the loop forever until we either dispose of the queue or we cancel the operation.